File: Algorithms/binary search/binary_search_recursive.py



def binary_search(array: list, target: float):
    
    
    if len(array) == 1:
        
        if array[0] == target:
            print("the given array included the given target")
        else:
            print("the given array not included the given target")
            
        return
          
            
    
    mid_index = (len(array)-1)//2
    
    if array[mid_index] == target:
        print("the given array included the given target")
        return
    
    if target > array[mid_index]:
        binary_search(array[mid_index+1:], target) 
    else:
        binary_search(array[:mid_index+1], target)












if __name__ == "__main__":
    
    print("\n\n ---------------- \n")
    print("a simple test:\n\n\n")
    
    
    # create a sorted array
    array = sorted([-2, 0, 55, 6, -11, -56, 34, 7, 3, 8, 99, 12, 123, 4, 3, -2, 1.34, 34.8, 53, 7])
    print("array:  ", array, '\n')
    
    

    print("search for 4 in array:  ", end='')
    binary_search(array, 4)
    
    print("search for -343 in array:  ", end='')
    binary_search(array, -343)
    
    
    print("\n\n ---------------- \n\n")

File: Algorithms/binary search/BinarySearchRecursive.java
import java.util.ArrayList;
import java.util.List;





/**
 * Use {@link BinarySearchRecursive#binarySearchRecursive(List<Integer>, int)} to 
 *  search for a value in your sorted list
 * 
 * 
 * @author Mohammad Mahdi Malmasi
 * @version 0.1.0
 */
public class BinarySearchRecursive
{
    /**
     * This method search for a value in your {@code List<Integer>} with recursive binary search algorithm
     * 
     * @param array : a sorted {@code List<Integer>}
     * @param target : value that you search for it
     */
    public static void binarySearchRecursive(List<Integer> array, int target)
    {
        if (array.size() < 2)
        {
            if (array.get(0) == target)
                System.out.println("the given array included the given target");
            else
                System.out.println("the given array not included the given target");


            return;
        }



        int midIndex = (array.size()-1) / 2;

        if (array.get(midIndex) == target)
        {
            System.out.println("the given array included the given target");
            return;
        }

        if (target > array.get(midIndex))
            binarySearchRecursive(array.subList(midIndex+1, array.size()), target);
        else
            binarySearchRecursive(array.subList(0, midIndex+1), target);
    }



    /**
     * A simple example for {@link BinarySearchRecursive#binarySearchRecursive(list<Integer, int)}
     *  
     * @param args
     */
    public static void main(String[] args) 
    {
        ArrayList<Integer> array = new ArrayList<>();
        array.add(-45); array.add(-12); array.add(-2); array.add(0); array.add(3); array.add(23); array.add(78);
        array.add(123); array.add(222); array.add(345); array.add(444); array.add(654); array.add(765);
        

        System.out.println("\n\n ---------------- \n");
        System.out.println("a simple test:\n");
        

        System.out.println("array: " + array.toString() + "\n");


        System.out.print("search for -12 in array: ");
        binarySearchRecursive(array, -12);
        System.out.print("search for 111 in array: ");
        binarySearchRecursive(array, 111);


        System.out.println("\n\n ---------------- \n\n");
    }
}